src/main/java/org/apache/maven/plugins/source/AbstractSourceJarMojo.java

```diff
@@ -303,15 +304,24 @@ protected void packageSources(List<MavenProject> theProjects) throws MojoExecuti
             }
 
             if (attach) {
+                boolean requiresAttach = true;
                 for (Artifact attachedArtifact : project.getAttachedArtifacts()) {
-                    if (isAlreadyAttached(attachedArtifact, project, getClassifier())) {
-                        getLog().error("We have duplicated artifacts attached.");
-                        throw new MojoExecutionException("Presumably you have configured maven-source-plugin "
-                                + "to execute twice in your build. You have to configure a classifier "
-                                + "for at least one of them.");
+                    Artifact previouslyAttached = getPreviouslyAttached(attachedArtifact, project, getClassifier());
+                    if (previouslyAttached != null) {
+                        if (!outputFile.equals(previouslyAttached.getFile())) {
```

Comment:
I think we must check the (current and "future") content is the same otherwise we hide a bug with such a logic - ie if between first and second run you add sources then the output artifact is likely wrong.

Reply:
too complex, without real benefit: before, `mvn install deploy` worked in any case
this notion of recording really the content each time attach is done is over-complexifying things: it will detect for example if build is not reproducible = not at all the intent

Reply:
Hmm, while I agree it is useless this means this PR too otherwise it defeats the enhancements to prevent to write twice the same artifact and corrupt a build so do we just close it?

Reply:
> it defeats the enhancements to prevent to write twice the same artifact and corrupt a build

wrong:
1. it avoid adding twice: see the `boolean requiresAttach`
2. it checks if 2 attaches try to attach real different files = different file names = where you really need to configure a classifier

as I wrote on the ML
I think it does t

> he right job at differentiating 2 cases that the initial MSOURCES-121 merged too aggressively:
> 1. do not fail when executed twice with the same output file
> 2. but fail when re-attach happens to another output file = another classifier has to be configured

Reply:
Not really, I understand this PR but you introduce a regression which is that you can attach twice the same artifact due to two executions so this "fix" is actually a regression.

`requiresAttach` is never needed until the build is broken - ie you build twice the source artifact which means you had a broken one until it is exactly the same content (why I originally proposed to test it since it must never happen in sane builds so cost is supposed to be 0).

`requiresAttach` is more a `ignoreAlreadyAttachedArtifact` but looses the "attached artifact was not wrongly used" by another mojo case (ie you enable to process a source jar which is not the one deployed).

The fix for the two tickets is quite easy and is on build side (it is already fixed on maven side):

* disable attach-sources phase which was embedded in maven 3.9 (or use maven 4 which does not have it anymore)
* use the same attach-sources than in 3.9 (jar-no-fork instead of jar)

So think we don't need to loose the check which was introduced in the plugin.

To make it more obvious I think https://issues.apache.org/jira/browse/MSOURCES-121 is right and shouldn't be reverted by this PR.

Reply:
> you can attach twice the same artifact due to two executions 
no, you can try (by running twice as `mvn install deploy`), but it won't attach the second time: then it won't try to deploy twice, then it won't fail if repository refuses to overwrite

you're right it could be called `ignoreAlreadyAttachedArtifact`

so what you mean is that running `mvn install deploy` is what you call "the build is broken"

I'll say it differently: you say MSOURCES-140 141 and 143 should be closed as either. "fixed" or "invalid"?

Reply:
I'd close it as invalid but think they hide multiple issues:

* maven core shouldn't define attach-sources even for release profile -> fixed in v4
* any build shouldn't attach twice the same artifact -> TBD in related projects (so "not an issue" on maven side)

Ambiguity is that point #2 comes from point #1 which is in maven core.
Ideally a real "maven" fix should be to drop the release profile in 3.x branch (no strong opinion on it) https://github.com/apache/maven/blob/maven-3.x-next/maven-model-builder/src/main/resources/org/apache/maven/model/pom-4.0.0.xml#L59.

So for source plugin: invalid, for maven 3.9: valid, for maven 4: not an issue.

Hope it makes sense.

Reply:
you are just ignoring the `mvn install deploy` case
and something fixed in Maven 4 deserves a workaround in maven-source-plugin 3

Reply:
install deploy falls in the same category, bug in mvn

Reply:
then do nothing? keep ignoring? keep "fixed in Maven 4"?
we're staying in circle being a pain in the ass of users

Reply:
Agree with Herve here.
We should be more users friendly here.
Let's simply introduce a new flag called `ignoreAlreadyAttachedArtifact`.
If the artifact is already attached and the flag true then simply accept to override the previous entry (but log a warn) otherwise just fail as the change introduces MSOURCES-121
At least users will have the choice.

Reply:
I'm ok to drop this PR in favor of a new boolean config parameter: who can code it, please?
I don't see how to code "override the previous entry"

Reply:
@hboutemy just replace it by documentation, if you reread the threads it is just people don't understanding maven superpom notion, we did our homework and fixed it now it is just a matter of managing the legacy.

@hboutemy @olamy I'm ok with `ignoreAlreadyAttachedArtifact` if it is not a `@Parameter` and if it is initialized with `mvn.version == 3` + a comment `must be false for maven 4 since see MSOURCES-121 and @github/#24`. I'm not ok to do positive enhancements then revert them because we are not able to explain what maven does or do a new maven release - literally this PR from my window.

Reply:
I added new IT for MSOURCES-140 and improved messages to clarify cases:

```
[ERROR] Artifact org.apache.maven.its.sources:jar-no-fork:java-source:sources:1.0-SNAPSHOT already attached to a file target/jar-no-fork-1.0-SNAPSHOT-sources.jar: attach to target/jar-no-fork-1.0-SNAPSHOT-secondary-sources.jar should be done with another classifier
```
if build is trying to attach 2 different files with same coordinates => failure and ask for configuring classifier

```
[INFO] Artifact org.apache.maven.its.sources:jar-no-fork:java-source:sources:1.0-SNAPSHOT already attached to target/jar-no-fork-1.0-SNAPSHOT-sources.jar: ignoring same re-attach (same artifact, same file)
```
if re-executing same goal configuration twice: just ignore the second time